Sources/Basics/Concurrency/ConcurrencyHelpers.swift

```diff
@@ -76,3 +77,87 @@ extension DispatchQueue {
+
+    // This implementation is adapted from the one in swift-build,
+    // modified to respect cancellation of the parent Task.
+    // https://github.com/swiftlang/swift-build/blob/054f2300ad83fd1633f1b50a06b82eea9e7c6901/Sources/SWBUtil/AsyncOperationQueue.swift#L13
+
+    typealias ID = UUID
+
+    private let concurrentTasks: Int
+    private var activeTasks: Int = 0
+    private var waitingTasks: [(ID, CheckedContinuation<Void, any Error>)] = []
+    private let waitingTasksLock = NSLock()
+
+    /// Creates an `AsyncOperationQueue` with a specified number of concurrent tasks.
+    /// - Parameter concurrentTasks: The maximum number of concurrent tasks that can be executed concurrently.
+    public init(concurrentTasks: Int) {
+        self.concurrentTasks = concurrentTasks
+    }
+
+    deinit {
+        waitingTasksLock.withLock {
+            if !waitingTasks.isEmpty {
+                preconditionFailure("Deallocated with waiting tasks")
+            }
+        }
+    }
+
+    /// Executes an asynchronous operation, ensuring that the number of concurrent tasks
+    // does not exceed the specified limit.
+    /// - Parameter operation: The asynchronous operation to execute.
+    /// - Returns: The result of the operation.
+    /// - Throws: An error thrown by the operation, or a `CancellationError` if the operation is cancelled.
+    public func withOperation<ReturnValue>(
+        _ operation: @Sendable () async throws -> sending ReturnValue
+    ) async throws -> ReturnValue {
+        try await waitIfNeeded()
+        defer { signalCompletion() }
+        return try await operation()
+    }
+
+    private func waitIfNeeded() async throws {
+        let shouldWait = waitingTasksLock.withLock {
+            let shouldWait = activeTasks >= concurrentTasks
+            activeTasks += 1
+            return shouldWait
+        }
+
+        if shouldWait {
+            let taskId = ID()
+
+            try await withTaskCancellationHandler {
+                try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, any Error>) in
+                    if !Task.isCancelled {
+                        waitingTasksLock.withLock {
+                            waitingTasks.append((taskId, continuation))
+                        }
+                    } else {
+                        continuation.resume(throwing: CancellationError())
+                    }
+                }
+            } onCancel: {
+                // If the parent task is cancelled then we need to manually handle resuming the
+                // continuation for the waiting task with a `CancellationError`.
+                self.waitingTasksLock.withLock {
+                    if let taskIndex = self.waitingTasks.firstIndex(where: { $0.0 == taskId }) {
+                        let task = self.waitingTasks.remove(at: taskIndex)
+                        task.1.resume(throwing: CancellationError())
```

Comment:
resuming continuations while a lock is held is dangerous and can lead to deadlocks. Note withTaskCancellationHandler's documentation:

> Cancellation handlers which acquire locks must take care to avoid deadlock. The cancellation handler may be invoked while holding internal locks associated with the task or other tasks. Other operations on the task, such as resuming a continuation, may acquire these same internal locks. Therefore, if a cancellation handler must acquire a lock, other code should not cancel tasks or resume continuations while holding that lock.

I think here you can return the continuation out of the lock and then resume it after releasing the lock.